src/Lime.Protocol.UnitTests/Network/ChannelBaseTests.cs

```diff
@@ -311,7 +311,33 @@ public async Task SendMessageAsync_TokenCancelled_DoNotCloseTransport()
             // Assert
             _transport.Verify(t => t.CloseAsync(It.IsAny<CancellationToken>()), Times.Never());
         }
-        
+
+        [Test]
+        [Category("SendMessageAsync")]
+        public async Task SendMessageAsync_LargeEnvelope_DoesNotCompleteTheBuffer()
+        {
+            // Arrange
+            var tcs = new TaskCompletionSource<Envelope>();
+            _transport
+                .Setup(t => t.SendAsync(It.IsAny<Envelope>(), It.IsAny<CancellationToken>()))
+                .Throws(new EnvelopeTooLargeException("too large envelope sent"));
+            var content = Dummy.CreateTextContent();
+            var message = Dummy.CreateMessage(content);
+            var target = GetTarget(SessionState.Established);
+
+            // Act
+            await target.SendMessageAndDelayAsync(message, CancellationToken.None);
+
+            // Assert - should resolve other messages
+            Assert.DoesNotThrowAsync(() => target.SendMessageAndDelayAsync(message, CancellationToken.None), "The send buffer is complete");
+
+            Assert.IsTrue((target).IsEstablished());
+            _transport.Verify(
```

Comment:
This last assert is not necessary, as this behavior (having the transport's `SendAsync` called) is irrelevant for the scenario you're testing. I presume the reason you're doing this is to make sure that the transport was really called; However, you should trust that it was since there are other tests that test that specific behavior. A single unit test should ideally test a single behavior/a single part of the contract.

Also, there's a negative consequence of testing implementation details that are not relevant to the specific test in hand, which is making the test fragile; We may, in the future, for whatever reason, decide to call the transport's `SendAsync` more than two times, without breaking the contract at all. This test would fail in that case, even though the fact that whether the transport's `SendAsync` is called, 2, 3 or 4 times for each call of the channel's `SendAsync` has nothing to do with the contract it is testing.

Reply:
The second assert is not necessary. It is being evauated true no matter what.

But the internal logic of the SenderChannel is blocking the path of a second call of the transport at a second execution.

I do not see why should I not invoke a verify at this test. All tests at this same region are verifing at least once the number of times the transport invoked the sendAsync. 

At the hypothetical scenario that we change the contract of the number of calls to the tranport per call of the Channel, you would have to change this entire file.

Reply:
![image](https://user-images.githubusercontent.com/22640498/128921840-7112a06b-3420-4254-b500-12711c6e85c9.png)
proving that the assert channel.IsEstablished  is not enough.

Reply:
> All tests at this same region are verifing at least once the number of times the transport invoked the sendAsync.

Yea, I just now saw that unfortunately all other tests are like this 😅

> The second assert is not necessary. It is being evauated true no matter what.

If you mean the one that asserts `IsEstablished`, that one is actually the important one, as it is effectively asserting what you wrote on the test name.

> At the hypothetical scenario that we change the contract of the number of calls to the tranport per call of the Channel, you would have to change this entire file.

And this is precisely why this test is fragile. This test is not responsible for testing this specific part of the contract. If that specific part of the contract was changed (how many calls to the transport's `SendAsync` are made), but the rest of the contract remained intact (including, for instance, the part of the contract that says that the channel remains open and connected when a too large envelope is sent) this test should _not_ break. 

Q: "But what if the channel isn't actually sending the calls to the Transport as I expect?" 
A: We should trust that there are other tests testing that. The same way that we don't assert that `System.Console.WriteLine` is really writing to the console or that `System.DateTime.Now` is really returning the correct date or that `System.Math.Sqrt(double)` is actually calling the C's stdlib extern function `sqrt` - why? because we assume that there are tests to test those cases.

Reply:
> ![image](https://user-images.githubusercontent.com/22640498/128921840-7112a06b-3420-4254-b500-12711c6e85c9.png)
> proving that the assert channel.IsEstablished is not enough.

Aaah I see. Makes sense now. In that case, what you want to do ideally is to test the state, instead of the interaction here. 
Suggestion: Mock the `SendAsync` method to capture the envelope it receives. Then check if the two messages you sent are contained in that list; Ideally the two messages should be different. Essentially you'll be doing a Spy here.

This way the test will still be depending a little bit on knowledge of implementation details (know the specific method of the transport that is called), but I think it's less fragile than the current one and helps to move in a better direction, serving as an example for whoever maintains this next.

Reply:
As we talked, let's look at this later.